libsolidity/lsp/LanguageServer.cpp

```diff
@@ -0,0 +1,966 @@
+		Json::Value item = Json::objectValue;
+		item["uri"] = toFileURI(_location.source->name());
+		item["range"] = toJsonRange(_location);
+		return item;
+	}
+
+	Json::Value toJson(LineColumn _pos)
+	{
+		Json::Value json = Json::objectValue;
+		json["line"] = max(_pos.line - 1, 0);
+		json["character"] = max(_pos.column - 1, 0);
+
+		return json;
+	}
+} // }}} end helpers
+
+LanguageServer::LanguageServer(Logger _logger, istream& _in, ostream& _out):
+	m_client{make_unique<JSONTransport>(_in, _out, _logger)},
+	m_handlers{
+		{"cancelRequest", [](auto, auto) {/*don't do anything for now, as we're synchronous*/}},
+		{"$/cancelRequest", [](auto, auto) {/*don't do anything for now, as we're synchronous*/}},
+		{"initialize", bind(&LanguageServer::handle_initialize, this, _1, _2)},
+		{"initialized", {} },
+		{"shutdown", [this](auto, auto) { m_shutdownRequested = true; }},
+		{"workspace/didChangeConfiguration", bind(&LanguageServer::handle_workspace_didChangeConfiguration, this, _1, _2)},
+		{"textDocument/didOpen", bind(&LanguageServer::handle_textDocument_didOpen, this, _1, _2)},
+		{"textDocument/didChange", bind(&LanguageServer::handle_textDocument_didChange, this, _1, _2)},
+		{"textDocument/didClose", [this](auto, Json::Value const& _args) {
+			documentClosed(extractPathFromFileURI(_args["textDocument"]["uri"].asString()).value());
+		}},
+		{"textDocument/definition", bind(&LanguageServer::handle_textDocument_definition, this, _1, _2)},
+		{"textDocument/documentHighlight", bind(&LanguageServer::handle_textDocument_highlight, this, _1, _2)},
+		{"textDocument/references", bind(&LanguageServer::handle_textDocument_references, this, _1, _2)},
+	},
+	m_logger{std::move(_logger)},
+	m_vfs()
+{
+}
+
+void LanguageServer::changeConfiguration(Json::Value const& _settings)
+{
+	if (_settings["evm"].isString())
+		if (auto const evmVersionOpt = EVMVersion::fromString(_settings["evm"].asString()); evmVersionOpt.has_value())
+			m_evmVersion = evmVersionOpt.value();
+
+	if (_settings["remapping"].isArray())
+	{
+		for (auto const& element: _settings["remapping"])
+		{
+			if (element.isString())
+			{
+				if (auto remappingOpt = CompilerStack::parseRemapping(element.asString()); remappingOpt.has_value())
+					m_remappings.emplace_back(move(remappingOpt.value()));
+				else
+					trace("Failed to parse remapping: '"s + element.asString() + "'");
+			}
+		}
+	}
+}
+
+void LanguageServer::documentContentUpdated(string const& _path, std::optional<int> _version, LineColumnRange _range, std::string const& _text)
+{
+	// TODO: all this info is actually unrelated to solidity/lsp specifically except knowing that
+	// the file has updated, so we can  abstract that away and only do the re-validation here.
+	auto file = m_vfs.find(_path);
+	if (!file)
+	{
+		log("LanguageServer: File to be modified not opened \"" + _path + "\"");
+		return;
+	}
+
+	if (_version.has_value())
+		file->setVersion(_version.value());
+
+	file->modify(_range, _text);
```

Comment:
I don't see why `m_vfs` can't be a `std::map<std::string, std::string>` (or even something internal to the compiler stack). This function - the most complicated part of VFS - could then be something like
```suggestion
	if (!m_vfs.count(_path))
	{
		log("LanguageServer: File to be modified not opened \"" + _path + "\"");
		return;
	}
	SourceLocation location = sourceLocationFromLineColumn(_path, _range);
	m_vfs[_path].replace(location.start, location.end, _text);
```

Reply:
After the code has been trimmed down to a minimum, we can think about a refactoring that avoids copying the strings all the time. Maybe it would be best to keep the strings owned inside the CompilerStack and just provide functions that return non-const references so we can modify them. Alternatively, we can copy them out and set single sources for the compiler, since changing some text inside the source requires copying anyway.

Reply:
Too bad I still think we need the version associated with a document state. I'm working something out. I was never fond with my current solution wrt. VFS myself either. :)